dev: include example RustRover run configurations, update nix-shell docs

sdht0 opened the pull request:
Running the equivalent of `bin/environmentd --reset` is not simple in RustRover. Include example run configuration in the repo for new developers to easily get started.

jkosh44 commented on .run/Run environmentd.run.xml:
```diff
@@ -0,0 +1,22 @@
+<component name="ProjectRunConfigurationManager">
+  <configuration default="false" name="Run environmentd" type="CargoCommandRunConfiguration" factoryName="Cargo Command">
+    <option name="buildProfileId" value="dev" />
+    <option name="command" value="run --bin environmentd -- --startup-log-filter=warn --system-parameter-default=log_filter=warn --internal-http-listen-addr=0.0.0.0:6878 --orchestrator=process --orchestrator-process-secrets-directory=$PROJECT_DIR$/mzdata/secrets --orchestrator-process-tcp-proxy-listen-addr=0.0.0.0 --orchestrator-process-prometheus-service-discovery-directory=$PROJECT_DIR$/mzdata/ ...
```
Is this identical to the params passes when running `bin/environmentd`?

sdht0 replied:
Ah thanks for making me check. I had left in some additional filters I use locally.

Update from a fresh cli run:
```
environmentd --internal-http-listen-addr=0.0.0.0:6878 --orchestrator=process --orchestrator-process-secrets-directory=/Users/siddharthasahu/Downloads/installations/materialize/mzdata/secrets --orchestrator-process-tcp-proxy-listen-addr=0.0.0.0 --orchestrator-process-prometheus-service-discovery-directory=/Users/siddharthasahu/Downloads/installations/materialize/mzdata/prometheus --orchestrator-process-scratch-directory=/Users/siddharthasahu/Downloads/installations/materialize/scratch --secrets-controller=local-file --persist-consensus-url=postgres://root@localhost:26257/materialize?options=--search_path=consensus --persist-blob-url=file:///Users/siddharthasahu/Downloads/installations/materialize/mzdata/persist/blob --timestamp-oracle-url=postgres://root@localhost:26257/materialize?options=--search_path=tsoracle --environment-id=local-az1-77fedbe0-a02c-479f-8252-398c66aae4f5-0 --bootstrap-role=materialize
```

jkosh44 reviewed (commented):
The contents all LGTM, but I'm not sure what our policy is on including editor specific files in our repo. Do we have any similar VS Code or other IDE files in the repo?

sdht0 commented:
> The contents all LGTM, but I'm not sure what our policy is on including editor specific files in our repo. Do we have any similar VS Code or other IDE files in the repo?

Ah that is a good question. Indeed I do not see any existing editor-specific config.

If not in a new root `.run` directory, hopefully we can atleast merge this with the configs inside `docs/developer/idea-configs` or `.config/idea`.

jkosh44 commented:
> > The contents all LGTM, but I'm not sure what our policy is on including editor specific files in our repo. Do we have any similar VS Code or other IDE files in the repo?
> 
> Ah that is a good question. Indeed I do not see any existing editor-specific config.
> 
> If not in a new root `.run` directory, hopefully we can atleast merge this with the configs inside `docs/developer/idea-configs` or `.config/idea`, so that we can have a reference config in one place, instead of each of us independently figuring out how to make it work.

I thought I remember reading in one of our style guides that we don't include IDE specific files in the main repo, but now I can't find that anywhere so it may have been my imagination. I'll leave it up to you whether you think this fits best somewhere in the repo or a notion doc or somewhere else.

antiguru commented:
My preference is not to include these files in the repo. They're both specific to the IDE and the user, and they need to be maintained as the interface to our binaries change. Having one canonical way to run a test setup greatly simplifies the burden on each engineer.

sdht0 commented:
> My preference is not to include these files in the repo. They're both specific to the IDE and the user, and they need to be maintained as the interface to our binaries change. Having one canonical way to run a test setup greatly simplifies the burden on each engineer.

Is there another suitable way we can make these config available to everyone?

The duplication in logic is indeed not ideal. However, anyone using RustRover to develop Materialize will end up having to create these configs and it'd be nice to be able to link to these in the docs.

antiguru commented:
We have some things in `/misc/editor/`.


> However, anyone using RustRover to develop Materialize will end up having to create these configs

Counterpoint: I'm using Rustrover only for `cargo check`, and terminal for `bin/environmentd`, and it works well for me.

sdht0 commented:
> We have some things in `/misc/editor/`.

Ah! This seems like an ideal location for the configs.

> > However, anyone using RustRover to develop Materialize will end up having to create these configs
> 
> Counterpoint: I'm using Rustrover only for `cargo check`, and terminal for `bin/environmentd`, and it works well for me.

Of course!

I'd miss some convenience features such as clicking paths in errors to jump to code directly and debugging in RustRover.

antiguru commented:
> debugging in RustRover

It's well-hidden, but there's Run > Attach profiler to process.

benesch commented:
+1 on including these in misc/editor/rustrover rather than .run, where it's more clear that these are files that you use at your own risk and not something that's officially maintained. (See misc/nix for a similar thing.)

For Visual Studio Code, we're able to use symlinks to that folder to good effect for folks who do want to take the risk: https://github.com/MaterializeInc/materialize/tree/main/misc/editor/vscode#installation.

antiguru commented on doc/developer/guide.md:
````diff
@@ -216,10 +216,17 @@ and needs to be installed separately, as described above. Also, IDEs will not be
 ```
 
+Note that on macOS, the `mzcompose` tests fail to run from within `nix-shell`,
+as our config does not yet set up [cross-compilation support](https://github.com/MaterializeInc/materialize/blob/main/misc/python/materialize/xcompile.py)
````
Use a relative link here and below.

sdht0 commented:
Instead of trying to emulate `--reset` from `bin/environmentd` inside RustRover, I added a flag to build and reset only in the Python script for use in the run config. Also de-duplicated the macos signature file. Hoping this minor addition to script is acceptable.

sdht0 commented on misc/python/materialize/cli/run.py:
```diff
@@ -244,6 +241,8 @@ def main() -> int:
                 )
                 paths.extend(p for p in scratch.glob("*"))
+                tmp = pathlib.Path(tempfile.gettempdir())
+                paths.extend(p for p in tmp.glob("environmentd-*"))
```
Highlighting this new cleanup inside the `/tmp` directory.

benesch replied:
This seems like it's only necessary because the RustRover configuration hardcodes a specific environment ID. That's mayyybe a fine approach but it's not really a supported configuration, so would rather not have run.py have special reset logic to account for that—too much action at a distance. Seems like something that should go in the RustRover-specific reset script instead.

sdht0 replied:
@benesch while I've reverted the changes, I do think this extra cleanup in `/tmp` should be part of the script. `--reset` does a full cleanup except for these files in `/tmp` and instead relies on unique environment ids. Across multiple `bin/environmentd --reset` runs, `/tmp` will continue to accrue multiple folders. Not that big a deal, but I think it would still be nice to explicitly mark all the places materialize stores data that might need a cleanup.

Happy to open another PR if you agree.

sdht0 commented on misc/python/materialize/cli/run.py:
```diff
@@ -506,12 +512,8 @@ def _macos_codesign(path: str) -> None:
-
-    command.append(temp.name)
+    macos_entitlements_data = MZ_ROOT / "misc" / "dist" / "macos-entitlements-data.xml"
+    command.append(macos_entitlements_data)
```
Doesn't need to be shared anymore, but still feels better than creating a temporary file on every run.

benesch replied:
Counterpoint: tempfiles are cheap, and the old approach kept the entitlement data nicely encapsulated inside of this script. I think it's kind of a wash, so fine to leave this in.

benesch reviewed (changes requested):
Only skimmed, but not psyched about adding another argument to `bin/environmentd` that intersects in a not immediately obvious way with the existing `--build-only` and `--reset` options. I'll take a closer look when I have a moment later today, but my preference would be to have the RustRover configuration recreate whatever logic from run.py that it needs, rather than trying to share logic. 

Separately, `bin/environmentd` really is the supported way to run environmentd locally.

benesch commented on doc/developer/guide.md:
```diff
@@ -570,13 +579,16 @@ If you are using Rust-Analyzer, you should configure it to conform to our
-Materialize employee, ask Nikhil Benesch on Slack for access to our corporate
-JetBrains license. If you're not yet sure you want to use RustRover, you can
-use the 30-day free trial.
+This is a good choice for developers who prefer the JetBrains ecosystem. This repo
```
```suggestion
This is a good choice for developers who prefer the JetBrains ecosystem. This folder
```

benesch commented on misc/dist/macos-entitlements-data.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
misc/dist is specifically for files that support our (now defunct) debian package—let's put this in misc/python/materialize/cli/run/macos-entitlements-data.xml instead.

benesch commented on misc/python/materialize/cli/run.py:
```diff
@@ -157,6 +150,10 @@ def main() -> int:
     )
     args = parser.parse_intermixed_args()
 
+    # `--build-and-reset-only` implies `--reset`
+    if args.build_and_reset_only:
+        args.reset = args.build_and_reset_only
```
The way `--build-only`, `--reset`, and `--build-and-reset-only` interact is now quite confusing.

I think you could instead make specifying the existing options (`--build-only --reset`) have the behavior you want/need for RustRover. Specifically: just move the `if args.build_only: return build_retcode` clauses to after the reset action happens. That will preserve the spirit of the existing options, have clear semantics, and IIUC give you the semantics you want for the RustRover config.

benesch commented on misc/editor/rustrover/run-environmentd.run.xml:
```diff
@@ -0,0 +1,30 @@
+<component name="ProjectRunConfigurationManager">
+  <configuration default="false" name="run-environmentd" type="CargoCommandRunConfiguration" factoryName="Cargo Command">
+    <option name="buildProfileId" value="dev" />
+    <option name="command" value="run --bin environmentd -- --internal-http-listen-addr=0.0.0.0:6878 --orchestrator=process --orchestrator-process-secrets-directory=$PROJECT_DIR$/mzdata/secrets --orchestrator-process-tcp-proxy-listen-addr=0.0.0.0 --orchestrator-process-prometheus-service-discovery-directory=$PROJECT_DIR$/mzdata/prometheus --orchestrator-process-scratch-directory=$PROJECT_DIR$/scr ...
```
This seems hard to maintain. As long as the blast radius of this approach is limited to misc/editor/rustrover that seems fine—you'll just have to update this as necessary for your workflow as run.py evolves.

An alternative to consider is adding a `--skip-build` option to bin/environmentd. That _is_ a bit of extra complexity for run.py, but it's complexity that I'd be willing to absorb to make these editor-specific configurations more maintainable. The `--skip-build` option would be mutually exclusive with the `--build-only` option and when specified would skip the Cargo/Bazel build step and instead expect the binaries to be already built. Then you'd just have to figure out how to adjust this run configuration to build the binaries, put them in the spot that bin/environmentd expects, and then invoke `bin/environmentd --skip-build`.

That way you'd automatically pick up new changes to bin/environmentd without needing to update your RustRover configuration.

antiguru commented:
You could just run `bin/environmentd` in RustRover directly instead of emulating the commands it'll run. I think that's a much better alternative than trying to build your own tooling. It'll break, and it'll cost you (or someone else using the tooling) time to figure out what's wrong and how to fix it.

sdht0 commented:
Since `bin/environmentd` is the supported way to run materialize, I've also moved to using this from RustRover instead of custom run configurations. There was some head-banging related to `nix-shell`, but I got there eventually and the included configuration works.
* Links to code line in panics or log messages are clickable.
* Debugging doesn't work. Since I've moved to `println`-based debugging, I've found myself using the IDE debugging less and less. Moreover, Moritz's suggestion to use `Run > Attach profiler to process` is a good workaround.

benesch reviewed (approved):
Nice, LGTM!

sdht0 commented:
TFTRs!
